KAFKA-3010; Include error in log when ack 0

ijuma opened the pull request:
I verified this by trying to produce to __consumer_offsets and the logged message looks like:

[2015-12-22 10:34:40,897] INFO [KafkaApi-0] Closing connection due to error during produce request with correlation id 1 from client id console-producer with ack=0
Topic and partition to exceptions: [__consumer_offsets,43] -> kafka.common.InvalidTopicException (kafka.server.KafkaApis)

ijuma committed:
Include error code when logging an error when ack = 0

ijuma committed:
Remove redundant and inaccurate comment

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -315,19 +315,23 @@ class KafkaApis(val requestChannel: RequestChannel,
-            topicAndPartition,
-            ErrorMapping.exceptionNameFor(status.error)))
+        if (status.error != ErrorMapping.NoError) {
           errorInResponse = true
```
This change means that we will log the "Closing connection" message even if the error is unknown. It seems to me that it's what we want, but I'd like someone more familiar with this code to confirm. cc @junrao

gwenshap replied:
We want to log the message even if error is unknown, but perhaps log the error code in that case? I'm not sure this should happen on the broker-side, but will be nice to have some information if it does.

ijuma replied:
We log the exception name which is UnknownException, which I think is more friendly than -1 (the error code). What do you think?

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -315,19 +315,23 @@ class KafkaApis(val requestChannel: RequestChannel,
           errorInResponse = true
+          // we only print warnings for known errors here; if it is unknown, it will cause
+          // an error message in the replica manager
+          if (status.error != ErrorMapping.UnknownCode) {
```
Perhaps we should just remove the if test and log UnknownException as well.

ijuma replied:
Sounds good. Many other requests have similarly looking code, should those be changed as well or should we only change it here?

junrao replied:
The two other usages are in debug logging too. It won't hurt to log the same thing in debug mode again. So, we can log UnknownExecption in those cases too.

ijuma replied:
Will do.

ijuma committed:
Merge remote-tracking branch 'apache/trunk' into kafka-3010-include-error-in-log-when-ack-0

* apache/trunk:
  KAFKA-3016: phase-1. A local store for join window
  KAFKA-3051 KAFKA-3048; Security config docs improvements
  KAFKA-2422: Allow copycat connector plugins to be aliased to simpler names
  KAFKA-3052; Broker properties get logged twice if acl enabled
  KAFKA-3055; Fix JsonConverter mangling the Schema in Connect
  KAFKA-2949; Make EndToEndAuthorizationTest replicated.
  MINOR: fixing typos in docs
  Minor: fix sasl.kerberos.service.name
  KAFKA-3043: Replace request.required.acks with acks in docs.
  MINOR: Improve document of MirrorMaker
  MINOR: Update to Gradle 2.10
  KAFKA-3002; Allow uppercase letters in hostnames
  MINOR: Fix a typos in comments
  KAFKA-3024: Remove old patch review tools
  KAFKA-3030: Remove unused scala dependencies
  MINOR: Tiny optimization to avoid mapping twice on the results.
  MINOR: Use "new line" delimeters in a portable way in a unit test
  KAFKA-2989; System tests should verify partitions consumed after rebalancing
  MINOR: Fix typo in documentation

ijuma committed:
Log `UnknownCode` errors at debug level in `KafkaApis`

Not worth excluding them since they are only logged at
debug level according to Jun.

ijuma commented:
@junrao I made the change you requested. Tests pass locally.

junrao commented:
Thanks for the patch. LGTM
